TestNG 7.4.0 throws an exception "sun.net.www.protocol.file.FileURLConnection cannot be cast to java.net.HttpURLConnection" when xml file contain "ENTITY  SYSTEM" grammer

### TestNG Version
> Note: only the latest version is supported
7.4.0
### Expected behavior
works as 7.1.0 

### Actual behavior
TestNG 7.4.0 throws an exception "sun.net.www.protocol.file.FileURLConnection cannot be cast to java.net.HttpURLConnection" when case xml file contain raw XML mechanism grammer like below:

```
<!ENTITY raw_listeners SYSTEM "testng/_listener/raw_listeners.xml">
        ]>
```

see java stack:

```
Connected to the target VM, address: '127.0.0.1:62526', transport: 'socket'
[TestNGContentHandler] [WARN] Failed to read [/E:/A_Project/A_Code/ose-test/jss_test/testng/_listener/raw_listeners.xml] from CLASSPATH. Attempting to read from [file:///E:/A_Project/A_Code/ose-test/jss_test/testng/_listener/raw_listeners.xml].
org.testng.TestNGException: java.lang.ClassCastException: sun.net.www.protocol.file.FileURLConnection cannot be cast to java.net.HttpURLConnection
	at org.testng.TestNG.parseSuite(TestNG.java:356)
	at org.testng.TestNG.initializeSuitesAndJarFile(TestNG.java:376)
	at com.intellij.rt.testng.IDEARemoteTestNG.run(IDEARemoteTestNG.java:39)
	at com.intellij.rt.testng.RemoteTestNGStarter.main(RemoteTestNGStarter.java:109)
Caused by: java.lang.ClassCastException: sun.net.www.protocol.file.FileURLConnection cannot be cast to java.net.HttpURLConnection
	at org.testng.xml.TestNGContentHandler.lambda$new$0(TestNGContentHandler.java:70)
	at org.testng.xml.TestNGContentHandler.resolveEntity(TestNGContentHandler.java:138)
	at com.sun.org.apache.xerces.internal.util.EntityResolverWrapper.resolveEntity(EntityResolverWrapper.java:110)
	at com.sun.org.apache.xerces.internal.impl.XMLEntityManager.resolveEntityAsPerStax(XMLEntityManager.java:997)
	at com.sun.org.apache.xerces.internal.impl.XMLEntityManager.startEntity(XMLEntityManager.java:1212)
	at com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl.scanEntityReference(XMLDocumentFragmentScannerImpl.java:1908)
	at com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl$FragmentContentDriver.next(XMLDocumentFragmentScannerImpl.java:3061)

```

### Is the issue reproductible on runner?

- [X ] Shell
- [X ] Maven
- [ ] Gradle
- [ ] Ant
- [ ] Eclipse
- [ ] IntelliJ
- [ ] NetBeans


### Test case sample
> Please, share the test case (as small as possible) which shows the issue
1. write a testng xml case file like this:
```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE suite SYSTEM "http://testng.org/testng-1.0.dtd" [
        <!ENTITY raw_listeners SYSTEM "testng/_listener/raw_listeners.xml">
        ]>
<suite name="ECBench Suite">
  <groups>
    <run>
      <exclude name = "known_bug"/>
      <exclude name = "disabled"/>
    </run>
  </groups>
  <test name="ecbench test">
    <classes>
      <class name="com.cloud.test.ecbench.ecmaster.Node.AddNodeTest" />
    </classes>
  </test> 
  &raw_listeners;
</suite>
```


and write another file which will be imported by using XML ENTITY grammer
```
<?xml version="1.0" encoding="UTF-8"?>

<listeners>
    <listener class-name="com.cloud.report.listener.TestNgReportListener"/>
</listeners>
```

2.  execute this case, and testng 7.4.0 will throw an exception "sun.net.www.protocol.file.FileURLConnection cannot be cast to java.net.HttpURLConnection"

but testng versions before 7.1.0 can recognize ENTITY in xml case files,

@krmahadevan See attached archive. Tried to do everything as simple as possible, so I went for the public static void main method and classpath in cmd.
[TestProject.zip](https://github.com/cbeust/testng/files/7736651/TestProject.zip)

Running with TestNG 6.9.11 everything is ok
`d:\TestProject\bin>java -cp ./;../lib/testng-6.9.11.jar;../lib/jcommander-1.78.jar Main`
produces the output(among other things)
`Total tests run: 1, Failures: 0, Skips: 0`

But running with TestNG 7.4.0 it crashes:
`d:\TestProject\bin>java -cp ./;../lib/testng-7.4.0.jar;../lib/jcommander-1.78.jar Main`
results in 
```
[TestNGContentHandler] [WARN] Failed to read [/d:/TestProject/bin/params.xml] from CLASSPATH. Attempting to read from [file:///d:/TestProject/bin/params.xml].
Exception in thread "main" org.testng.TestNGException: java.lang.ClassCastException: class sun.net.www.protocol.file.FileURLConnection cannot be cast to class java.net.HttpURLConnection (sun.net.www.protocol.file.FileURLConnection and java.net.HttpURLConnection are in module java.base of loader 'bootstrap')
        at org.testng.TestNG.parseSuite(TestNG.java:356)
        at org.testng.TestNG.initializeSuitesAndJarFile(TestNG.java:376)
        at org.testng.TestNG.initializeEverything(TestNG.java:1011)
        at org.testng.TestNG.run(TestNG.java:1023)
        at Main.main(Main.java:10)
Caused by: java.lang.ClassCastException: class sun.net.www.protocol.file.FileURLConnection cannot be cast to class java.net.HttpURLConnection (sun.net.www.protocol.file.FileURLConnection and java.net.HttpURLConnection are in module java.base of loader 'bootstrap')
```
openjdk 14.0.2 2020-07-14
OpenJDK Runtime Environment (build 14.0.2+12-46)
OS: Windows 10
I'm pretty sure that the jdk version is not relevant, I also tested with 8, but not with the attached files.

--- a/CHANGES.txt
+++ b/CHANGES.txt
@@ -1,4 +1,5 @@
 Current
+Fixed: GITHUB-2501: TestNG 7.4.0 throws an exception "sun.net.www.protocol.file.FileURLConnection cannot be cast to java.net.HttpURLConnection" when xml file contain "ENTITY SYSTEM" grammer (Krishnan Mahadevan)
 Fixed: GITHUB-2685: TestInvoker should clear Thread.interrupted flag before calling ITestListeners (Roman Morskyi)
 Fixed: GITHUB-2684: AfterGroups config annotation does not consider retries for tests (Roman Morskyi)
 Fixed: GITHUB-2689: Yaml parser: implement loadClasses flag (Dzmitry Sankouski)

--- a/testng-core/src/main/java/org/testng/xml/TestNGContentHandler.java
+++ b/testng-core/src/main/java/org/testng/xml/TestNGContentHandler.java
@@ -11,6 +11,7 @@ import java.net.MalformedURLException;
 import java.net.URI;
 import java.net.URISyntaxException;
 import java.net.URL;
+import java.net.URLConnection;
 import java.net.URLDecoder;
 import java.nio.charset.StandardCharsets;
 import java.util.ArrayList;
@@ -70,17 +71,22 @@ public class TestNGContentHandler extends DefaultHandler {
                   "Failed to read [%s] from CLASSPATH. " + "Attempting to read from [%s].",
                   url.getPath(), systemId);
           Logger.getLogger(getClass()).warn(msg);
-          HttpURLConnection conn = (HttpURLConnection) url.openConnection();
+          URLConnection urlConnection = url.openConnection();
+          if (urlConnection instanceof HttpURLConnection) {
+            HttpURLConnection conn = (HttpURLConnection) urlConnection;
 
-          int status = conn.getResponseCode();
-          if ((status == HttpURLConnection.HTTP_MOVED_TEMP
-              || status == HttpURLConnection.HTTP_MOVED_PERM
-              || status == HttpURLConnection.HTTP_SEE_OTHER)) {
+            int status = conn.getResponseCode();
+            if ((status == HttpURLConnection.HTTP_MOVED_TEMP
+                || status == HttpURLConnection.HTTP_MOVED_PERM
+                || status == HttpURLConnection.HTTP_SEE_OTHER)) {
 
-            String newUrl = conn.getHeaderField("Location");
-            conn = (HttpURLConnection) new URL(newUrl).openConnection();
+              String newUrl = conn.getHeaderField("Location");
+              conn = (HttpURLConnection) new URL(newUrl).openConnection();
+            }
+            stream = conn.getInputStream();
+          } else {
+            stream = urlConnection.getInputStream();
           }
-          stream = conn.getInputStream();
         }
         return new InputSource(
             Objects.requireNonNull(stream, "Failed to load DTD from " + systemId));

--- a/testng-core/src/test/java/test/xml/TestNGContentHandlerTest.java
+++ b/testng-core/src/test/java/test/xml/TestNGContentHandlerTest.java
@@ -1,12 +1,17 @@
 package test.xml;
 
+import static org.assertj.core.api.Assertions.assertThat;
+
 import java.io.FileInputStream;
+import java.io.IOException;
 import java.util.List;
 import org.testng.Assert;
 import org.testng.annotations.Test;
 import org.testng.xml.SuiteXmlParser;
 import org.testng.xml.TestNGContentHandler;
+import org.testng.xml.XmlClass;
 import org.testng.xml.XmlInclude;
+import org.xml.sax.SAXException;
 import test.SimpleBaseTest;
 
 public class TestNGContentHandlerTest extends SimpleBaseTest {
@@ -20,6 +25,16 @@ public class TestNGContentHandlerTest extends SimpleBaseTest {
         handler.getSuite().getTests().get(0).getXmlClasses().get(0).getIncludedMethods();
     String desc = includes.get(0).getDescription();
     Assert.assertEquals("simple-description", desc);
+  }
+
+  @Test(description = "GITHUB-2501")
+  public void ensureAppropriateConnectionObjectsAreUsed() throws IOException, SAXException {
+    String xml = getPathToResource("xml/issue2501/2501.xml");
+    SuiteXmlParser parser = new SuiteXmlParser();
+    TestNGContentHandler handler = new TestNGContentHandler(xml, false);
+    parser.parse(new FileInputStream(xml), handler);
+    XmlClass xmlClass = handler.getSuite().getTests().get(0).getXmlClasses().get(0);
+    assertThat(xmlClass.getSupportClass()).isEqualTo(test.xml.issue2501.TestClassSample.class);
   }
 
   public static class LocalTestClass {

--- a/testng-core/src/test/java/test/xml/issue2501/TestClassSample.java
+++ b/testng-core/src/test/java/test/xml/issue2501/TestClassSample.java
@@ -1,0 +1,9 @@
+package test.xml.issue2501;
+
+import org.testng.annotations.Test;
+
+public class TestClassSample {
+
+  @Test
+  public void testMethod() {}
+}

--- a/testng-core/src/test/resources/xml/issue2501/2501.xml
+++ b/testng-core/src/test/resources/xml/issue2501/2501.xml
@@ -1,0 +1,17 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE suite SYSTEM "http://testng.org/testng-1.0.dtd" [
+  <!ENTITY params SYSTEM "src/test/resources/xml/issue2501/params.xml">
+  ]>
+
+<suite name="VerifyTests" parallel="false" thread-count="1" preserve-order="true">
+
+  &params;
+
+  <parameter name="param" value="Value"/>
+
+  <test name="TestName">
+    <classes>
+      <class name="test.xml.issue2501.TestClassSample"/>
+    </classes>
+  </test>
+</suite>

--- a/testng-core/src/test/resources/xml/issue2501/params.xml
+++ b/testng-core/src/test/resources/xml/issue2501/params.xml
@@ -1,0 +1,3 @@
+<?xml version="1.0" encoding="UTF-8"?>
+
+<parameter name="paramImported" value="ValueFromImportedFile" />

